tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+// Author: Ethan Uppal
+
+//! This library serves as a rust version of calyx-py. The main builder API is
+//! via [`Program`], from where you can create [`Component`]s. Most `struct`s
+//! here are [`CalyxWriter`]s, so you can easily obtain their calyx
+//! representations (although, of course, that of [`Program`] is arguably the
+//! most useful). Macros have been developed for creating cells, groups, and
+//! control; they are the *intended* way of constructing these elements.
+
+use std::{
+    cell::RefCell,
+    fmt::{self, Display, Write},
+    rc::Rc,
+};
+
+/// Shorthand for the `Rc<RefCell<>>` pattern.
+pub type RRC<T> = Rc<RefCell<T>>;
+
+/// Convenience constructor for [`RRC`].
+fn rrc<T>(value: T) -> RRC<T> {
+    Rc::new(RefCell::new(value))
+}
+
+struct IndentFormatter<'a, 'b: 'a> {
+    current_indent: String,
+    add_indent: usize,
+    last_was_newline: bool,
+    formatter: &'a mut fmt::Formatter<'b>,
+}
+
+impl<'a, 'b: 'a> IndentFormatter<'a, 'b> {
+    /// Constructs a new formatter managing indents of `indent` spaces in the
+    /// wrapped formatter `formatter`.
+    pub fn new(formatter: &'a mut fmt::Formatter<'b>, indent: usize) -> Self {
+        Self {
+            current_indent: String::new(),
+            add_indent: indent,
+            last_was_newline: false,
+            formatter,
+        }
+    }
+
+    /// Adds a level of indentation.
+    pub fn push(&mut self) {
+        for _ in 0..self.add_indent {
+            self.current_indent.push(' ');
+        }
+    }
+
+    /// Removes a level of indentation.
+    pub fn pop(&mut self) {
+        for _ in 0..self.add_indent {
+            self.current_indent.pop();
+        }
+    }
+}
+
+impl fmt::Write for IndentFormatter<'_, '_> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        for c in s.chars() {
+            if self.last_was_newline {
+                self.formatter.write_str(&self.current_indent)?;
+            }
+            self.formatter.write_char(c)?;
+            self.last_was_newline = c == '\n';
+        }
+        Ok(())
+    }
+}
+
+/// A location in the source code for source element `element` and optionally a
+/// unique identifier `id` for that element. For example,
+/// ```
+/// // COMPONENT END: main
+/// ```
+/// would be a marker for the "component" element and with a unique identifier
+/// of "main".
+struct Marker {
+    element: String,
+    id: Option<String>,
+}
+
+impl Marker {
+    /// A marker for an `element`.
+    #[allow(dead_code)]
+    pub fn general<S: ToString>(element: S) -> Marker {
+        Marker {
+            element: element.to_string(),
+            id: None,
+        }
+    }
+
+    /// A marker for a unique `element` identified by `id`.
+    pub fn unique<S: ToString, T: ToString>(element: S, id: T) -> Marker {
```

Comment:
Good to know!

Reply:
This requires using `String::from(x.as_ref())` instead of `x.to_string()` for a given `x: AsRef<str>` -- is that better to have?

Reply:
You can use `x.as_ref().to_string()` which should get you what you want. The difference is that this excludes random structures which implement `Display` which would otherwise be allowed